[windows] Identify SDR WCG screens as non HDR-capable

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

(description modified since first version of the PR, which attempted to do the same thing but using poorly documented Win32 API and educated guesses)

Starting with Windows 11 22H2, Windows supports a new type of screen, "Wide Color Gamut" (WCG) and interop methods were added to allow retrieving the type of screen (including WCG) using a winrt API.
That API also works with Xbox so a new unified code path was created, and the legacy methods for Win32 and UWP remain for now, but are isolated for easy removal eventually.

The WCG screens caused a problem with Win11 because the Win32 API used until now was confusing them with an HDR capable screen, and as a result Kodi would attempt to switch Windows HDR on/off when playing videos, with bad visual results. At this time there is no documented way to detect WCG screens with the Win32 API.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Helps with issue #24615, misdetection of WCG screen as HDR.

Some screens have wide color support and an on/off hdr software switch.
No problem with the on position, detected as expected by Kodi as HDR capable screen and used as such.
In the off position however, Microsoft added support for non-hdr wcg screens in Windows 11 and recent AMD drivers seem to take advantage of that, resulting in incorrect HDR detection by Kodi due to the way it's reported by the Win32 API.

Attempting to use the WCG capabilities with standard dynamic range is beyond the scope of v21 but at least the screen could be detected as SDR, which allows tonemapping.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Win11/Intel works fine, Xbox works
The Win32 detection still works on Windows 8 and 10.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Avoid incorrect switch to non-working HDR mode when the screen is in SDR mode.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai commented on xbmc/platform/win32/WIN32Util.cpp:
```diff
@@ -1424,7 +1424,7 @@ HDR_STATUS CWIN32Util::GetWindowsHDRStatus()
                 advancedColorEnabled = true;
 
-              if (getColorInfo.advancedColorSupported)
+              if (getColorInfo.advancedColorSupported && !getColorInfo.wideColorEnforced)
```
This is a ugly hack that mostly will break and cause issues in all others systems. Please not do that.

`getColorInfo.advancedColorSupported` by itself signals that display _supports_ HDR  (enabled or not enabled) you can not add other conditions here.

HDR may be supported but not enabled in Windows as you know (and even not enabled needs to be detected as supported).

`getColorInfo.advancedColorEnabled` by itself signals that HDR is _enabled_ this by logic includes that is supported.

thexai replied:
> AdvancedColorSupported. This indicates that your installed display driver and display support HDR10.
> 
> AdvancedColorEnabled. This indicates that HDR is currently turned on for your display.

https://support.microsoft.com/en-us/windows/hdr-settings-in-windows-2d767185-38ec-7fdc-6f97-bbc6c5ef24e6

thexai replied:
Also original issue seems is not working tone mapping (dark picture) but is for unrelated reason. In all systems tone mapping should work and works with HDR_OFF:

// HDR_STATUS::HDR_OFF
advancedColorSupported = true;
advancedColorEnabled = false;

I mean the required condition is HDR_OFF only (not HDR_UNSUPPORTED)

=> Display supports HDR but is OFF then use tone mapping.

CrystalP replied:
The OP had two separate issues, screen HDR capabilities and broken dxva processor conversion (the reason for broken tonemapping). This deals only with the first.

I'd rather trust API documentation than description of user settings, which can have some processing on top of the API.

SDR/wide color support is new in Windows 11, see "Windows 11, version 22H2: SDR displays with Advanced Color" , "This topic assumes that your app is targeting Windows 10, version 2004 (or later) for HDR displays, and the Windows 11, version 22H2 release (or later) for SDR displays." "Option 1: AdvancedColorInfo" in https://learn.microsoft.com/en-us/windows/win32/direct3darticles/high-dynamic-range to understand what AdvancedColor means to Microsoft.

The Win32 structure has unclear documentation, only in the SDK header.
See related WinRT API, https://learn.microsoft.com/en-us/uwp/api/windows.graphics.display.advancedcolorkind?view=winrt-22621 and https://learn.microsoft.com/en-us/uwp/api/windows.graphics.display.advancedcolorinfo?view=winrt-22621

AdvancedColor doesn't mean only HDR. Right from the start they planned the WideColorGamut value as a different advanced color value, with SDR transfer and BT2020 colors.

The Win32 structure exposes bits so it was not possible to represent the same way as WinRT 0, 1, 2. Instead:

advancedColorSupported = 0 means SDR
advancedColorSupported = 1 and wideColorEnforced = 1 means SDR with wide colors
advancedColorSupported = 1 and wideColorEnforced = 0 means PQ transfer + wide colors (the usual HDR)

This is supported by the log files in the ticket for my special builds that logged all the bits of DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO.

I agree that for clarity I should have changed the Kodi variable name to "hdrSupported" but the condition is good. I'll update.

I know it's kind of risky and wouldn't mind merging this in early v22 to have plenty of feedback instead.

An alternative is to switch Kodi build in order to allow calling directly the native WinRT API, there are a few others that could be interesting (override blue light filter, dynamically detect dark mode, there will be more in the future)

thexai replied:
> advancedColorSupported = 0 means SDR
> advancedColorSupported = 1 and wideColorEnforced = 1 means SDR with wide colors
> advancedColorSupported = 1 and wideColorEnforced = 0 means PQ transfer + wide colors (the usual HDR)

In my desktop PC with Nvidia and SDR only monitor is:

![color-info](https://github.com/xbmc/xbmc/assets/58434170/e2e5f66b-2278-40db-b1c8-2b5314d8a3ef)

Then is not clear...

advancedColorSupported = 0 and wideColorEnforced = 1 means ?

thexai replied:
> I know it's kind of risky and wouldn't mind merging this in early v22 to have plenty of feedback instead.

I agree

CrystalP replied:
Without proper documentation it's difficult. It seems that usage of wideColorEnforced is new in Win11.
It's always 0 for me in Win 10 and all vendors.

the rule "advancedColorSupported = 0 means SDR" still works for your config.

I would have liked to know the results of the WinRT API for the OP (and you) but didn't find a quick way to do it.

thexai replied:
My concern is that a display that supports both HDR and WCG probably _may_ have:

advancedColorSupported = 1
wideColorEnforced = 1

With current PR will broke HDR --> Detected as not HDR capable then Kodi is unable to toggle Windows HDR on

CrystalP replied:
For OP both flags =1 mean SDR with WCG (HDR setting of the screen turned off)
His screen reports advancedColorSupported = 1 wideColorEnforced = 0 for "true" HDR, the same as me, and you check your system too.
But it's sample size of 1.

CrystalP commented:
Freshened up the code, renamed a variable for clarity.
If this doesn't work , would result in some users saying their screen is not detected as HDR anymore.

Doc from MS hasn't improved in the meantime.

In https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ne-wingdi-displayconfig_device_info_type there are additional values that I didn't notice before
DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO_2
DISPLAYCONFIG_DEVICE_INFO_SET_HDR_STATE
DISPLAYCONFIG_DEVICE_INFO_SET_WCG_STATE

They seem interesting but the Windows 11 SDK 10.0.22621.0 bundled with Visual Studio doesn't have them or the structs they need to work. Not sure where to get a more recent SDK?

An alternative is to change Kodi to call WinRT. Not sure what it takes and consequences. What do you think?

thexai reviewed (dismissed):
Approved but I think this will cause regressions that you will have to take care of later.

For same reason we should not backport.

CrystalP commented:
The safe way is the WinRT API, but do you have an idea what it would take?

CrystalP committed:
[windows] Identify SDR WCG screens as non HDR-capable

Win 11 22621 (22H2) supports WCG screens and also has some interop functions with Win32 to use
the new WinRT API for the detection.
Unify the Win11 desktop and Xbox paths, kept the older code for older desktop and UWP desktop.
The older code can also be used as fallback in case the new method doesn't work.

CrystalP commented:
Pushed some changes using the WinRT API, made possible by a couple interop methods added in Win11 22H2 to retrieve a DisplayInformation object for a Win32 monitor or window.
I attempted some unification of the UWP and Win32 winrt code but it's still a work in progress and the #ifdefs are ugly...

I had limited access to a Win11 system and could check that the IsAdvancedColorKindAvailable() API works to detect screens capable of HDR even with HDR off in Windows (Intel GPU).

I don't know what happens on Xbox or UWP on Win11 though, it would help a lot if you could give it a try.

Either IsAdvancedColorKindAvailable() works or we still need the original detection code based on DisplayManager(), that I kept around for now.

CrystalP commented:
Windows build bot failure is likely due to a requirement for a more recent Windows SDK (PR builds fine locally with Windows 11 SDK 10.0.22621.0)

thexai commented on xbmc/platform/win10/CMakeLists.txt:
```diff
@@ -5,6 +5,7 @@ set(SOURCES CPUInfoWin10.cpp
             Win10App.cpp
             MessagePrinter.cpp
             PlatformWin10.cpp
+            WinRtUtil.cpp
```
Being a common code for Win32 and UWP, it must be in the `win32` path. I know the name is not much appropriate and somewhat confusing but the `win10` path only contains UWP exclusive code (not used in desktop path).

WinRtUtil name is fine, only move to xbmc/platform/win32

CrystalP replied:
ok, done, as well as the other comments.

thexai commented on xbmc/platform/win10/WinRtUtil.h:
```diff
@@ -0,0 +1,20 @@
+class CWinRtUtil
+{
+public:
+  CWinRtUtil(void);
```
```suggestion
  CWinRtUtil();
```
and define on .cpp file or change here to `CWinRtUtil() = default;`

Same for  `~CWinRtUtil()` and `virtual` not seems necessary here

CrystalP replied:
I was lazy and c/p from WIN32Util.h but actually there is no need for a public constructor or destructor at this point. Could be added later when needed.

thexai commented on xbmc/rendering/dx/DirectXHelper.h:
```diff
@@ -202,18 +202,4 @@ namespace DX
-    }
-  }
-}
-#endif
```
This change is necessary?

CrystalP replied:
Yes, that was a crutch until now to use some winrt types in Win32 without including the proper headers. Now that the headers are needed, that code caused type redefinition conflicts.

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -641,6 +642,7 @@ void DX::DeviceResources::ResizeBuffers()
   else // Otherwise, create a new one using the same adapter as the existing Direct3D device.
   {
     HDR_STATUS hdrStatus = CWIN32Util::GetWindowsHDRStatus();
+    CWinRtUtil::GetWindowsHDRStatus();
```
This is necessary?   It's not called from inside CWIN32Util::GetWindowsHDRStatus(); when supported?

Not used return value?

CrystalP replied:
that's a leftover from an earlier iteration of the PR, will be removed.

thexai reviewed (commented):
Only initial review, not tested yet...

thexai commented:
Tested on Xbox Series S and confirmed this code is working to detect HDR off but supported:

```c++
    // TODO: Combine with desktop case if works for Xbox
    if (colorInfo.IsAdvancedColorKindAvailable(AdvancedColorKind::HighDynamicRange))
      return HDR_STATUS::HDR_OFF;
    else
      return HDR_STATUS::HDR_UNSUPPORTED;
```

This is very good a allows use same HDR detection code for x64 desktop (Windows 11 only) and Xbox

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -23,6 +23,7 @@
 
 #include "platform/win32/CharsetConverter.h"
 #include "platform/win32/WIN32Util.h"
+#include "platform/win32/WinRtUtil.h"
```
After remove `CWinRtUtil::GetWindowsHDRStatus()` call this header is not needed

CrystalP replied:
correct

CrystalP commented on xbmc/platform/win32/WinRtUtil.cpp:
```diff
@@ -0,0 +1,201 @@
+#ifdef TARGET_WINDOWS_DESKTOP
+#include <Windows.Graphics.Display.Interop.h>
+
+#elif defined(TARGET_WINDOWS_STORE)
```
ok

CrystalP commented:
Added a commit to address yesterday's comments.
It's very good news that the Xbox can use the same detection as Win 11 desktop. Thanks for the testing.

thexai commented on xbmc/platform/win32/WIN32Util.cpp:
```diff
@@ -1469,6 +1429,35 @@ HDR_STATUS CWIN32Util::GetWindowsHDRStatus()
+
+#ifdef TARGET_WINDOWS_STORE
+
+  if (CSysInfo::GetWindowsDeviceFamily() == CSysInfo::Xbox &&
```
It's only a personal preference but since Visual Studio editor grey out not used code path (desktop/UWP) I prefer no extra blank lines inside `#ifdef` code blocks.

i.e. this instead of current:

![after](https://github.com/xbmc/xbmc/assets/58434170/bf5f7cd8-982a-467a-bc97-fc4cf7db3015)
 
here and similar others

CrystalP replied:
I don't really care one way or the other. Updated.

thexai reviewed (commented):
Tested also on Intel NUC with Windows 11 23H2 and no issues found.

thexai commented on xbmc/platform/win32/WIN32Util.cpp:
```diff
@@ -1347,53 +1348,13 @@ HDR_STATUS CWIN32Util::ToggleWindowsHDR(DXGI_MODE_DESC& modeDesc)
 
-HDR_STATUS CWIN32Util::GetWindowsHDRStatus()
+#ifdef TARGET_WINDOWS_DESKTOP
+static HDR_STATUS GetWindowsHDRStatusWin32()
```
For consistency with WinRtUtil maybe include `GetWindowsHDRStatusWin32` method in `CWIN32Util` class and return HDR_STATUS::HDR_UNKNOWN if its called from target store.

I mean:

```c++
HDR_STATUS CWIN32Util::GetWindowsHDRStatusWin32()
{
  HDR_STATUS status = HDR_STATUS::HDR_UNKNOWN;

#ifdef TARGET_WINDOWS_DESKTOP
...
#endif

  return status;
}
```

CrystalP replied:
I wondered about the same thing. Done, as a private method, there is no good reason to call it from outside CWIN32Util.

CrystalP committed:
[Windows] Update build requirements with Windows 11 SDK 10.0.22621.0

CrystalP commented:
Push a commit with requested changes and small refactor of advanced color kind logging.
Also a commit with Windows 11 SDK bump requirement for build, which will not be squashed with the rest at the end

fuzzard commented:
Jenkins build this please

CrystalP commented:
Squashed all commits except the SDK version requirement bump.
Thanks to the SDK bump on the builder the PR builds successfully.

thexai commented on xbmc/platform/win32/WinRtUtil.cpp:
```diff
@@ -0,0 +1,203 @@
+static void LogAdvancedColorInfo(AdvancedColorInfo colorInfo)
+{
+  std::string availableAdvColorKind;
+  std::array<AdvancedColorKind, 3> kinds{AdvancedColorKind::StandardDynamicRange,
```
```suggestion
  constexpr std::array<AdvancedColorKind, 3> kinds{AdvancedColorKind::StandardDynamicRange,
```

CrystalP replied:
done, made static as well.

thexai commented on xbmc/platform/win32/WinRtUtil.cpp:
```diff
@@ -0,0 +1,203 @@
+                                         AdvancedColorKind::WideColorGamut,
+                                         AdvancedColorKind::HighDynamicRange};
+
+  for (auto kind : kinds)
```
```suggestion
  for (const auto& kind : kinds)
```

CrystalP replied:
ok for the principle. It's an enum so no performance gained,

thexai commented:
Windows SDK also should be increased here:

https://github.com/xbmc/xbmc/blob/3d14ee659bc0a9f431ee094487a781d8ef0ad892/cmake/scripts/windows/ArchSetup.cmake#L2

and:

https://github.com/xbmc/xbmc/blob/3d14ee659bc0a9f431ee094487a781d8ef0ad892/cmake/scripts/windowsstore/ArchSetup.cmake#L2

CrystalP commented:
Thanks for finding the cmake SDK version check, my text search missed it.
Updated, as well as the rest of the comments.
